File: code/platform/src/main/java/org/fbme/ide/platform/editor/PlatformEditorFactory.java
package org.fbme.ide.platform.editor;

import jetbrains.mps.ide.editor.NodeEditorFactory;
import jetbrains.mps.ide.editor.NodeEditorFactoryBase;
import jetbrains.mps.internal.collections.runtime.IWhereFilter;
import jetbrains.mps.internal.collections.runtime.Sequence;
import jetbrains.mps.lang.smodel.generator.smodelAdapter.SLinkOperations;
import jetbrains.mps.lang.smodel.generator.smodelAdapter.SNodeOperations;
import jetbrains.mps.openapi.editor.Editor;
import jetbrains.mps.project.MPSProject;
import jetbrains.mps.smodel.adapter.structure.MetaAdapterFactory;
import jetbrains.mps.smodel.structure.ExtensionPoint;
import org.fbme.ide.iec61499.editor.ProjectEditorSpecs;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.jetbrains.mps.openapi.language.SConcept;
import org.jetbrains.mps.openapi.language.SReferenceLink;
import org.jetbrains.mps.openapi.model.SNode;

import java.util.List;
import java.util.stream.Collectors;

public class PlatformEditorFactory extends NodeEditorFactoryBase {

    private final MPSProject myProject;

    public PlatformEditorFactory(MPSProject project) {
        myProject = project;
    }

    @Override
    public boolean canCreate(@NotNull NodeEditorFactory.Context context) {
        final SNode node = context.getNode();
        return EditorProjectionControllerRegistry.getInstance()
                .getFactories().stream()
                .anyMatch(it -> it.isApplicable(node));
    }

    @Nullable
    @Override
    public Editor create(@NotNull NodeEditorFactory.Context context) {
        return new HeaderedNodeEditor(context.getNode(), myProject);
    }

    @Override
    public SNode getBaseNode(@NotNull SNode node) {
        return ProjectEditorSpecs.getInstance(myProject).getSpecTarget(node);
    }
}

File: code/platform/src/main/kotlin/org/fbme/ide/platform/editor/PlatformEditorFactory.kt
package org.fbme.ide.platform.editor

import com.intellij.openapi.project.Project
import jetbrains.mps.ide.editor.NodeEditorFactory
import jetbrains.mps.ide.editor.NodeEditorFactoryBase
import jetbrains.mps.openapi.editor.Editor
import jetbrains.mps.project.MPSProject
import org.fbme.ide.iec61499.editor.ProjectEditorSpecs
import org.fbme.ide.iec61499.repository.PlatformElement
import org.fbme.ide.iec61499.repository.PlatformRepositoryProvider
import org.jetbrains.mps.openapi.model.SNode

class PlatformEditorFactory(project: Project) : NodeEditorFactoryBase() {
    private val mpsProject = project.getComponent(MPSProject::class.java)

    override fun canCreate(context: NodeEditorFactory.Context): Boolean {
        val node = context.node
        val element = runCatching { PlatformRepositoryProvider.getInstance(mpsProject).getAdapter(node, PlatformElement::class.java) }.getOrNull()
                ?: return false
        return EditorProjectionControllerProvider.EP_NAME
            .extensions
            .any { it.isApplicable(element) }
    }

    override fun create(context: NodeEditorFactory.Context): Editor {
        return HeaderedNodeEditor(context.node, mpsProject)
    }

    override fun getBaseNode(node: SNode): SNode {
        return ProjectEditorSpecs.getInstance(mpsProject).getSpecTarget(node)
    }
}
